from_cpython/Objects/iterobject.c

```diff
@@ -75,7 +66,7 @@ calliter_next(calliterobject *it)
                 Py_CLEAR(it->it_sentinel);
             }
         }
-        else if (PyErr_ExceptionMatches(PyExc_StopIteration)) {
+        else if (PyErr_ExceptionMatches(PyExc_StopIteration) || PyErr_ExceptionMatches(PyExc_IndexError)) {
```

Comment:
Hm that does feel super fishy. I don't remember what prompted this but I'll take another look.